neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -137,130 +133,173 @@ def __init__(self,
-        # to (batch_size * beam_size)
-        # feedables have shape [batch, ...]
-        feedables = nest.map_structure(
-            lambda x: expand_to_beam(x, 0, self.beam_size),
-            dec_ls.feedables)
-        # histories have shape [len, batch, ...]
-        histories = nest.map_structure(
-            lambda x: expand_to_beam(x, 1, self.beam_size),
-            dec_ls.histories)
-
-        # We add input_symbol during token_ids initialization
-        # for simpler beam_body implementation.
-        self._bs_output = SearchStepOutput(
-            scores=tf.zeros(
-                shape=[self.batch_size, self.beam_size],
-                dtype=tf.float32,
-                name="beam_scores"),
-            token_ids=tf.reshape(
-                feedables.input_symbol,
-                [1, self.batch_size, self.beam_size],
-                name="beam_tokens"))
-
-        dec_ls = dec_ls._replace(
+        """Construct the initial loop state for the beam search decoder.
+
+        During the construction, the body function of the underlying decoder
+        is called once to retrieve the initial log probabilities of the first
+        token.
+
+        The values are initialized as follows:
+
+        - ``search_state``
+            - ``logprob_sum`` - For each sentence in batch, logprob sum of the
+              first hypothesis in the beam is set to zero while the others are
+              set to negative infinity.
+            - ``prev_logprobs`` - This is the softmax over the logits from the
+              initial decoder step.
+            - ``lengths`` - All zeros.
+            - ``finshed`` - All false.
+
+        - ``search_results``
+            - ``scores`` - A (batch, beam)-sized tensor of zeros.
+            - ``token_ids`` - A (1, batch, beam)-sized tensor filled with
+              indices of decoder-specific initial input symbols (usually start
+              symbol IDs).
+
+        - ``decoder_loop_state`` - The loop state of the underlying
+            autoregressive decoder, as returned from the initial call to the
+            body function.
+
+        Returns:
+            A populated ``BeamSearchLoopState`` structure.
+        """
+        # Get the initial loop state of the underlying decoder. Then, expand
+        # the tensors from the loop state to (batch * beam) and inject them
+        # back into the decoder loop state.
+
+        dec_init_ls = self.parent_decoder.get_initial_loop_state()
+
+        feedables = tf.contrib.framework.nest.map_structure(
+            self.expand_to_beam, dec_init_ls.feedables)
+        histories = tf.contrib.framework.nest.map_structure(
+            lambda x: self.expand_to_beam(x, dim=1), dec_init_ls.histories)
+
+        dec_init_ls = dec_init_ls._replace(
             feedables=feedables,
             histories=histories,
-            constants=tf.constant(0))
+            constants=tf.constant(0))  # TODO(@varisd) what is this?!
```

Comment:
The decoding crashes when constants contains an empty array. This is a workaround.

Reply:
But why don't we just copy/expand_to_beam whatever was in the decoder constants?